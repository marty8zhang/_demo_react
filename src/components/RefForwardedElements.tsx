import React, {
  ForwardedRef,
  PropsWithChildren,
  ReactNode,
} from 'react';

type Props = {
  forwardedRef: ForwardedRef<HTMLParagraphElement>,
}

class RefForwardingAwareP extends React.Component<Props> {
  render(): ReactNode {
    const { children, forwardedRef } = this.props;
    return (
      <p ref={forwardedRef}>{children}</p>
    );
  }
}

/*
 * `ref` only exists when you define a component with `React.forwardRef()`. Regular function or
 * class components donâ€™t receive the `ref` argument, and `ref` is not available in `props` either.
 * This is why in `RefForwardingAwareP` below we have to use a different name to pass down
 * the `ref` as a part of `props`.
 */
const RefForwardedP = React.forwardRef(
  (props: PropsWithChildren<{children?: ReactNode}>, ref: ForwardedRef<HTMLParagraphElement>) => (
    <RefForwardingAwareP forwardedRef={ref}>{props.children}</RefForwardingAwareP>
  ),
);

const RefForwardedH2 = React.forwardRef(
  (props: PropsWithChildren<{children?: ReactNode}>, ref: ForwardedRef<HTMLHeadingElement>) => (
    <h2 ref={ref}>{props.children}</h2>
  ),
);

export { RefForwardedP, RefForwardedH2 };
