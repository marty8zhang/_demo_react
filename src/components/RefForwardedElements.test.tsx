import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { RefForwardedH2, RefForwardedP } from './RefForwardedElements';

let container: HTMLDivElement | null = null;
const h2Ref = React.createRef<HTMLHeadingElement>();
const pRef = React.createRef<HTMLParagraphElement>();

it('matches with the H2 inline snapshot', () => {
  act(() => {
    render(
      <RefForwardedH2 ref={h2Ref}>This is a referred H2.</RefForwardedH2>,
      container,
    );
  });
  expect(h2Ref.current?.outerHTML).toMatchInlineSnapshot(
    '"<h2>This is a referred H2.</h2>"',
  );
});

it('matches with the paragraph snapshot', () => {
  act(() => {
    render(
      <RefForwardedP ref={pRef}>This is a referred paragraph.</RefForwardedP>,
      container,
    );
  });
  expect(pRef.current?.outerHTML).toMatchSnapshot();
});

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  if (!container) {
    return;
  }

  unmountComponentAtNode(container);
  container.remove();
  container = null;
});
