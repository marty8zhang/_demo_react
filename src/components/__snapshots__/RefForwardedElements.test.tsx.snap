// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches with the paragraph snapshot 1`] = `"<p>This is a referred paragraph.</p>"`;
